blockchain/treasury.go

```diff
@@ -0,0 +1,846 @@
+			// This should not happen.
+			trsyLog.Errorf("  checkTSpendExists: block not found "+
+				"%v tspend %v", v, tspend)
+			continue
+		}
+
+		if prevNode.Ancestor(node.height) != node {
+			trsyLog.Errorf("  checkTSpendExists: not ancestor "+
+				"block %v tspend %v", v, tspend)
+			continue
+		}
+		trsyLog.Errorf("  checkTSpendExists: is ancestor "+
+			"block %v tspend %v", v, tspend)
+		return fmt.Errorf("tspend has already been mined on this "+
+			"chain %v", tspend)
+	}
+
+	return nil
+}
+
+// CheckTSpendExists verifies that the provided TSpend has not been mined in a
+// block on the chain of the provided block hash.
+func (b *BlockChain) CheckTSpendExists(prevHash, tspend chainhash.Hash) error {
+	prevNode := b.index.LookupNode(&prevHash)
+	if prevNode == nil {
+		return nil
+	}
+	return b.checkTSpendExists(prevNode, tspend)
+}
+
+// getVotes returns yes and no votes for the provided hash.
+func getVotes(votes []stake.TreasuryVoteTuple, hash *chainhash.Hash) (yes int, no int, abstain int) {
+	if votes == nil {
+		return
+	}
+
+	for _, v := range votes {
+		if !hash.IsEqual(&v.Hash) {
+			continue
+		}
+
+		switch v.Vote {
+		case stake.TreasuryVoteYes:
+			yes++
+		case stake.TreasuryVoteNo:
+			no++
+		case stake.TreasuryVoteAbstain:
+			abstain++
+		default:
+			// Can't happen.
+			trsyLog.Criticalf("getVotes: invalid vote 0x%v", v.Vote)
+		}
+	}
+
+	return
+}
+
+// tspendVotes is a structure that contains a treasury vote tally for a given
+// window.
+type tspendVotes struct {
+	start   uint32 // Start block
+	end     uint32 // End block
+	yes     int    // Yes vote tally
+	no      int    // No vote tally
+	abstain int    // Abstain vote tally
+}
+
+// tSpendCountVotes returns the vote vote tally for a given tspend.
+// NOTE: This function must only be called on a TVI.
+func (b *BlockChain) tSpendCountVotes(prevNode *blockNode, tspend *dcrutil.Tx) (*tspendVotes, error) {
+	trsyLog.Tracef("tSpendCountVotes: processing tspend %v", tspend.Hash())
+
+	var (
+		t   tspendVotes
+		err error
+	)
+
+	expiry := tspend.MsgTx().Expiry
+	t.start, err = standalone.CalculateTSpendWindowStart(expiry,
+		b.chainParams.TreasuryVoteInterval,
+		b.chainParams.TreasuryVoteIntervalMultiplier)
+	if err != nil {
+		return nil, err
+	}
+	t.end, err = standalone.CalculateTSpendWindowEnd(expiry,
+		b.chainParams.TreasuryVoteInterval)
+	if err != nil {
+		return nil, err
+	}
+
+	nextHeight := prevNode.height + 1
+	trsyLog.Tracef("  tSpendCountVotes: nextHeight %v start %v expiry %v",
+		nextHeight, t.start, t.end)
+
+	// Ensure tspend is within the window.
+	if !standalone.InsideTSpendWindow(nextHeight,
+		expiry, b.chainParams.TreasuryVoteInterval,
+		b.chainParams.TreasuryVoteIntervalMultiplier) {
+		err = fmt.Errorf("tspend outside of window: nextHeight %v "+
+			"start %v expiry %v", nextHeight, t.start, expiry)
+		return nil, err
+	}
+
+	// Walk prevNode back to the start of the window and count votes.
+	node := prevNode
+	if node == nil {
```

Comment:
If this is the case, then the above will crash for accessing a nil node.

Reply:
This was a workaround for a condition that can no longer occur.